File: app/src/main/java/com/zack/enderplan/presenter/AboutPresenter.java
package com.zack.enderplan.presenter;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;

import com.zack.enderplan.R;
import com.zack.enderplan.util.CommonUtil;
import com.zack.enderplan.util.ResourceUtil;
import com.zack.enderplan.util.SystemUtil;
import com.zack.enderplan.view.adapter.AboutPagerAdapter;
import com.zack.enderplan.view.contract.AboutViewContract;
import com.zack.enderplan.view.fragment.AboutFragment;
import com.zack.enderplan.view.fragment.ThanksFragment;
import com.zack.enderplan.view.fragment.ProblemsFragment;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class AboutPresenter extends BasePresenter implements SensorEventListener {

    private static final int APP_BAR_STATE_EXPANDED = 1;
    private static final int APP_BAR_STATE_INTERMEDIATE = 0;
    private static final int APP_BAR_STATE_COLLAPSED = -1;

    private AboutViewContract mAboutViewContract;
    private AboutPagerAdapter mAboutPagerAdapter;
    private int mAppBarMaxRange;
    private float mLastHeaderAlpha = 1f;
    private int mAppBarState = APP_BAR_STATE_EXPANDED;
    private SensorManager mSensorManager;
    private float[] mGravityValues;
    private float[] mGeomagneticValues;
    private float[] mRotationValues;
    private float[] mOrientationValues;
    private float mMaxTranslation;

    @Inject
    AboutPresenter(AboutViewContract aboutViewContract, FragmentManager fragmentManager, SensorManager sensorManager) {
        mAboutViewContract = aboutViewContract;

        mAboutPagerAdapter = new AboutPagerAdapter(fragmentManager, getAboutPages());

        mSensorManager = sensorManager;
        mGravityValues = new float[3];
        mGeomagneticValues = new float[3];
        mRotationValues = new float[9];
        mOrientationValues = new float[3];
        mMaxTranslation = CommonUtil.convertDpToPx(12);
    }

    @Override
    public void attach() {
        mAboutViewContract.showInitialView(SystemUtil.getVersionName(), mAboutPagerAdapter);
    }

    @Override
    public void detach() {
        mAboutViewContract = null;
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if (event.sensor.getType() == Sensor.TYPE_ACCELEROMETER) {
            mGravityValues = event.values;
        }
        if (event.sensor.getType() == Sensor.TYPE_MAGNETIC_FIELD) {
            mGeomagneticValues = event.values;
        }
        SensorManager.getRotationMatrix(mRotationValues, null, mGravityValues, mGeomagneticValues);
        SensorManager.getOrientation(mRotationValues, mOrientationValues);

        float translationX = -mOrientationValues[2] * 30;
        float translationY = mOrientationValues[1] * 30;

        mAboutViewContract.translateViewWhenIncline(
                Math.abs(translationX) <= mMaxTranslation,
                translationX,
                Math.abs(translationY) <= mMaxTranslation,
                translationY
        );
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {

    }

    public void notifyPreDrawingAppBar(int appBarMaxRange) {
        mAppBarMaxRange = appBarMaxRange;
    }

    public void notifyAppBarScrolled(int offset) {

        if (mAppBarMaxRange == 0) return;

        int absOffset = Math.abs(offset);
        float headerAlpha = 1f - absOffset * 1.3f / mAppBarMaxRange;
        if (headerAlpha < 0) headerAlpha = 0;

        if ((headerAlpha == 0 || mLastHeaderAlpha == 0) && headerAlpha != mLastHeaderAlpha) {
            mAboutViewContract.onAppBarScrolledToCriticalPoint(headerAlpha == 0 ? ResourceUtil.getString(R.string.title_activity_about) : " ");
            mLastHeaderAlpha = headerAlpha;
        }

        mAboutViewContract.onAppBarScrolled(headerAlpha);

        if (absOffset == 0) {
            mAppBarState = APP_BAR_STATE_EXPANDED;
        } else if (absOffset == mAppBarMaxRange) {
            mAppBarState = APP_BAR_STATE_COLLAPSED;
        } else {
            mAppBarState = APP_BAR_STATE_INTERMEDIATE;
        }
    }

    public void notifyRegisteringSensorListener() {
        mSensorManager.registerListener(this, mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER), SensorManager.SENSOR_DELAY_UI);
        mSensorManager.registerListener(this, mSensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD), SensorManager.SENSOR_DELAY_UI);
    }

    public void notifyUnregisteringSensorListener() {
        mSensorManager.unregisterListener(this);
    }

    public void notifyResetingViewTranslation() {
        mAboutViewContract.resetViewTranslation();
    }

    public void notifyBackPressed() {
        if (mAppBarState == APP_BAR_STATE_EXPANDED) {
            mAboutViewContract.pressBack();
        } else {
            mAboutViewContract.backToTop();
        }
    }

    private List<Fragment> getAboutPages() {
        List<Fragment> fragmentList = new ArrayList<>();
        fragmentList.add(new AboutFragment());
        fragmentList.add(new ThanksFragment());
        fragmentList.add(new ProblemsFragment());
        return fragmentList;
    }
}

File: app/src/main/kotlin/me/imzack/app/end/presenter/AboutPresenter.kt
package me.imzack.app.end.presenter

import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.support.v4.app.FragmentManager
import me.imzack.app.end.R
import me.imzack.app.end.common.Constant
import me.imzack.app.end.util.CommonUtil
import me.imzack.app.end.util.ResourceUtil
import me.imzack.app.end.util.SystemUtil
import me.imzack.app.end.view.adapter.AboutPagerAdapter
import me.imzack.app.end.view.contract.AboutViewContract
import me.imzack.app.end.view.fragment.AboutFragment
import me.imzack.app.end.view.fragment.ProblemsFragment
import me.imzack.app.end.view.fragment.ThanksFragment
import javax.inject.Inject

class AboutPresenter @Inject constructor(
        private var mAboutViewContract: AboutViewContract?,
        fragmentManager: FragmentManager,
        private val mSensorManager: SensorManager
) : BasePresenter(), SensorEventListener {

    // 主构造函数的参数可以在init块中使用，也可以在类中声明的属性初始化器中使用
    private val mAboutPagerAdapter = AboutPagerAdapter(fragmentManager, listOf(AboutFragment(), ThanksFragment(), ProblemsFragment()))
    private var mAppBarMaxRange = 0
    private var mLastHeaderAlpha = 1f
    private var mAppBarState = Constant.APP_BAR_STATE_EXPANDED
    private var mGravityValues = FloatArray(3)
    private var mGeomagneticValues = FloatArray(3)
    private val mRotationValues = FloatArray(9)
    private val mOrientationValues = FloatArray(3)
    private val mMaxTranslation = CommonUtil.convertDpToPx(12).toFloat()

    override fun attach() {
        mAboutViewContract!!.showInitialView(SystemUtil.versionName, mAboutPagerAdapter)
    }

    override fun detach() {
        mAboutViewContract = null
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (event.sensor.type == Sensor.TYPE_ACCELEROMETER) {
            mGravityValues = event.values
        }
        if (event.sensor.type == Sensor.TYPE_MAGNETIC_FIELD) {
            mGeomagneticValues = event.values
        }
        SensorManager.getRotationMatrix(mRotationValues, null, mGravityValues, mGeomagneticValues)
        SensorManager.getOrientation(mRotationValues, mOrientationValues)

        val translationX = -mOrientationValues[2] * 30
        val translationY = mOrientationValues[1] * 30

        mAboutViewContract!!.translateViewWhenIncline(
                Math.abs(translationX) <= mMaxTranslation,
                translationX,
                Math.abs(translationY) <= mMaxTranslation,
                translationY
        )
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {

    }

    fun notifyPreDrawingAppBar(appBarMaxRange: Int) {
        mAppBarMaxRange = appBarMaxRange
    }

    fun notifyAppBarScrolled(offset: Int) {

        if (mAppBarMaxRange == 0) return

        val absOffset = Math.abs(offset)
        var headerAlpha = 1f - absOffset * 1.3f / mAppBarMaxRange
        if (headerAlpha < 0) headerAlpha = 0f

        if ((headerAlpha == 0f || mLastHeaderAlpha == 0f) && headerAlpha != mLastHeaderAlpha) {
            mAboutViewContract!!.onAppBarScrolledToCriticalPoint(if (headerAlpha == 0f) ResourceUtil.getString(R.string.title_activity_about) else " ")
            mLastHeaderAlpha = headerAlpha
        }

        mAboutViewContract!!.onAppBarScrolled(headerAlpha)

        mAppBarState = when (absOffset) {
            0 -> Constant.APP_BAR_STATE_EXPANDED
            mAppBarMaxRange -> Constant.APP_BAR_STATE_COLLAPSED
            else -> Constant.APP_BAR_STATE_INTERMEDIATE
        }
    }

    fun notifyRegisteringSensorListener() {
        mSensorManager.registerListener(this, mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER), SensorManager.SENSOR_DELAY_UI)
        mSensorManager.registerListener(this, mSensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD), SensorManager.SENSOR_DELAY_UI)
    }

    fun notifyUnregisteringSensorListener() {
        mSensorManager.unregisterListener(this)
    }

    fun notifyResetingViewTranslation() {
        mAboutViewContract!!.resetViewTranslation()
    }

    fun notifyBackPressed() {
        if (mAppBarState == Constant.APP_BAR_STATE_EXPANDED) {
            mAboutViewContract!!.pressBack()
        } else {
            mAboutViewContract!!.backToTop()
        }
    }
}
